### ISSUE TITLE
how the loss is calculated for timedistributed layer

### ISSUE THREAD

**Issue by joefang66** [keras-team/keras#8055]
Hi, 

here is a question about how the loss is calculated for timedistributed layer.

the model is shown below:

model = Sequential()  
model.add(GRU(hidden_neurons, input_dim=in_out_neurons, return_sequences=True))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(in_out_neurons)))
model.add(Activation("linear"))  
model.compile(loss="categorical_crossentropy", optimizer="rmsprop") 
print "Model compiled."

the shape of input X and output Y is (100,40,20), (100,40,2)

100 is batch size, 40 is timestep, 20 is input dim, 2 is output dim.

I am wondering how the loss is calculated. Does the loss function calculate the loss for each time step, or calculate the loss for whole time step (add the loss of each time step, and do BP)?

Thank you very much.

**Comment by philayres** [comment]
I had the same question, open on the Google group for a while: https://groups.google.com/forum/#!topic/keras-users/3aGqqMFBe6E 

>  have a simple sequential model using TimeDistributed(Dense...) after an LSTM layer. I am training on time series data in sequences of up to 100 time steps.

> I am trying to confirm whether the standard loss functions do actually calculate loss across all the time steps. Looking at the code it looks like the loss may just be calculated on a single time step, but that could be just my poor understanding.
> 
> When running my code loss seems to drop very slowly, indicating a problem.
> 
> Does anybody have good insight into the way the loss functions work when time series data is used with TimeDistributed as the output?

I believe this backs up the original question. Any reasons for this?

**Comment by joefang66** [comment]
Hi Phil,

I consult a senior PhD student in my university, he said the loss of each
timedistributed step will be added together finally. I am not sure whether
he is correct, but I have no other ways to get the answer.

I have post this question on github. Hope someone will answer it in the
future.

Thanks a lot.

Bests,
Joe

2017-10-05 19:05 GMT+08:00 Phil Ayres <notifications@github.com>:

**Comment by ylmeng** [comment]
I also want to know. My application has a TimeDistributed output layer, and I want the loss of each time slice to be independently treated (not a "batch updating" at the end).

**Comment by DomHudson** [comment]
Likewise, I am also interested in this. Presumably with a time-distributed dense output layer all timesteps will contribute to the loss?

**Comment by ylmeng** [comment]
^^ I believe they all contribute, however it is unclear to me whether the errors are back-propagated in each time step, or an average error is back-propagated over all timesteps (suppose batch size is 1 here). The effects can be very different especially when the timesteps are long.

**Comment by philayres** [comment]
I have received some great feedback on my original stack overflow question: https://stackoverflow.com/questions/45765136/loss-calculations-for-lstm-timedistributeddense-layers 

I think it may answer this question, although I'm still digesting the response.
